pkg/common-controller/snapshot_metrics.go

```diff
@@ -0,0 +1,77 @@
+package common_controller
+
+import (
+	"fmt"
+
+	crdv1 "github.com/kubernetes-csi/external-snapshotter/client/v3/apis/volumesnapshot/v1beta1"
+	"github.com/kubernetes-csi/external-snapshotter/v3/pkg/metrics"
+	"github.com/kubernetes-csi/external-snapshotter/v3/pkg/utils"
+	klog "k8s.io/klog/v2"
+)
+
+const (
+	createSnapshotOperationName = "CreateSnapshot"
+	snapshottingOperationName   = "Snapshotting"
+	deleteSnapshotOperationName = "DeleteSnapshot"
+
+	dynamicSnapshotType        = "dynamic"
+	preProvisionedSnapshotType = "pre-provisioned"
+)
+
+// snapshotKind represents which kind of snapshot a metric is
+type snapshotType string
+
+// SnapshotOperationStatus represents the status for a snapshot controller operation
+type SnapshotOperationStatus struct {
+	Status string
+	Error  string
+}
+
+// NewSnapshotOperationStatus returns a new SnapshotOperationStatus
+func NewSnapshotOperationStatus(err error) SnapshotOperationStatus {
+	if err != nil {
+		return SnapshotOperationStatus{
+			Status: "failed",
+			Error:  err.Error(),
+		}
+	}
+
+	return SnapshotOperationStatus{
+		Status: "success",
+	}
+}
+
+func (sos SnapshotOperationStatus) String() string {
```

Comment:
this implementation will likely create a LOT of time series of latency metrics (one for each error message) which is problematic. see:
https://prometheus.io/docs/practices/naming/#labels

the status string is a label of the latency metrics, we definitely need to limit the cardinality 
https://github.com/kubernetes-csi/external-snapshotter/blob/master/pkg/metrics/metrics.go#L134

Reply:
Added a type `snapshotStatusType` with four types as per the design doc:
1. Success
2. Invalid request
3. Controller error
4. Storage system error

However, I couldn't find any examples of storage-system-error? Wouldn't that mainly be in the snapshotter sidecar?

Reply:
For storage system error, are we able to get csi error code?

Reply:
We could check the VolumeSnapshot.Status.Error, but it seems like that error is used for both storage-system and controller errors. It formats the message as a string, I believe losing the csi error code.
https://github.com/kubernetes-csi/external-snapshotter/blob/23b415b6aaa7e0eba402bc3984ae2b726b101a80/pkg/sidecar-controller/snapshot_controller.go#L141-L156

So I'm not sure we can get the csi error code from that error, unless you are talking about checking in another spot.